include/vinecopulib/vinecop/implementation/class.ipp

```diff
@@ -1151,19 +1197,28 @@ Vinecop::rosenblatt(const Eigen::MatrixXd& u, const size_t num_threads) const
   }
 
   // go back to original order
-  auto U_vine = u;
+  Eigen::MatrixXd U(n, 2 * d);
   for (size_t j = 0; j < d; j++) {
-    U_vine.col(j) = hfunc2.col(inverse_order[j]);
+    U.col(j) = hfunc2.col(inverse_order[j]);
   }
 
-  return U_vine.array().min(1 - 1e-10).max(1e-10);
+  if (randomize_discrete && is_discrete()) {
+    for (size_t j = 0; j < d; j++) {
+      U.col(d + j) = var_types_[j] == "d" ? hfunc2_sub.col(inverse_order[j]) 
+                                          : hfunc2.col(inverse_order[j]);
+    }
+    auto R = tools_stats::simulate_uniform(u.rows(), d, true, seeds);
+    U.leftCols(d) = U.leftCols(d).array() * R.array() +
+      U.rightCols(d).array() * (1 - R.array());
+  } 
+  return U.leftCols(d).array().min(1 - 1e-10).max(1e-10);
 }
 
 //! @brief Evaluates the inverse Rosenblatt transform.
 //!
 //! @details The inverse Rosenblatt transform can be used for simulation: the
 //! function applied to independent uniform variates resembles simulated
-//! data from the vine copula model. Only works for continous models.
+//! data from the vine copula model. 
```

Comment:
Wait, but why did we restrict it to continuous models before given how minor the changes below are?

Reply:
Because forward `rosenblatt()` didn't work for discrete variables. Wasn't the greatest decision in hindsight.